templates/extras/simp.md

```diff
@@ -1,111 +1,265 @@
+
+
+Note that some `simp` lemmas have preconditions. For example, a
+theorem about cancelling a factor on both sides of an equation would
+only be valid under the hypothesis that the factor is non-zero. If `h`
+is a proof of hypothesis `P` and `P → A = B` is a `simp` lemma then
+`simp [h]` will replace `A`'s with `B`'s in the goal. The fact that
+`simp` works with preconditions is the reason it is called a
+*conditional* term rewriting system.
+
+## Simp-normal form
+
+There are sometimes several ways to say the same thing. For example,
+if `n : ℕ` then the hypotheses `n ≠ 0`, `0 ≠ n`, `n > 0`, `0 < n`, `1
+≤ n` and `n ≥ 1` are all ways to say the same thing. This can be
+problematic for rewriting systems like the simplifier. The reason is
+that the simplifier works up to *syntactic equality*. If the
+simplifier is working on a term `T`, and `A = B` is a `simp` lemma,
+and there is a subterm `A'` in `T` which is logically equal or
+equivalent to `A`, but not literally the same string of characters as
+`A`, then the simplifier will in general not replace it with
+`B`. Similarly, if nonzeroness of `n` (stated in one way) is a
+precondition in a `simp` lemma of the form `A = B`, and `h` is a proof
+of nonzeroness of `n` stated in a different way, then `simp [h]` might
+not replace `A`'s with `B`'s.
+
+A bad solution to this would be to write essentially the same lemma
+many times, so each possible invariant is covered. Clearly this is not
+ideal. The way this issue is dealt with in `mathlib` is to fix once
+and for all a so-called "simp normal form" for a way of expressing
+something such that `n ≠ 0` above, and then trying to stick to this
+variant when stating lemmas in Lean. For example, most lemmas which
+assume a natural is non-zero do it via the hypothesis `0 < n`.
+
+In general, if you are writing a lemma, you should know the "normal
+form" way to express the ideas in the lemma. If you are writing a
+lemma about a definition you made yourself, you might want to make
+your own design decision about the "normal form" way of expressing any
+idea which can be naturally expressed in more than one way.
+
+An example of a simp-normal form is a way of expressing nonemptiness
+of a subset of a type.  If `α : Type` and `s : set α` then
+nonemptiness of `s` can be expressed as `s.nonempty` and as `s ≠ ∅`.
+In mathlib an effort is made to stick to `s.nonempty` as the normal
+form.
+
+Another example: if `s : finset α` and `a : α` then `a ∈ s` is defined
+to mean `a ∈ (s : set α)`. The simp normal form for this idea is `a ∈
+s`, and moreover the `simp` lemma
+
+```
+@[simp] lemma mem_coe {a : α} {s : finset α} : a ∈ (s : set α) ↔ a ∈ s := ...
+```
+
+will attempt to turn any occurrences of `a ∈ (s : set α)` into the
+correct normal form.
+
+Because the simplifier works from the inside out, simplifying arguments
+of a function before simplifying the function, a `simp` lemma should
+always have all of its arguments already in simp-normal form. Mathlib's
```

Comment:
The issue is with the (arguments of the) LHS of the simp lemma, not the arguments to the lemma itself. For instance, if you have @[simp] lemma fg (cond : P) : f (g 0) = 0 but g 0 already simplifies to something else, fg will never fire. I read "arguments of fg" as referring to hypotheses like cond, but it doesn't matter if cond is in simp-normal form.
 
```suggestion
the arguments of the left hand side of a simp lemma 
should already be in simp-normal form. Mathlib's
```
?
You could add an example if you want.

Reply:
Thanks. It's clear that I'm not really the best person to be writing this page, however I am very much motivated to get something better than what we currently have because students can ask direct questions such as "how can I learn about `simp`?". Thanks all of you for the help you've given.